clang/lib/Sema/SemaDecl.cpp

```diff
@@ -4057,6 +4057,10 @@ bool Sema::MergeFunctionDecl(FunctionDecl *New, NamedDecl *&OldD, Scope *S,
                                                          NewQType))
       return MergeCompatibleFunctionDecls(New, Old, S, MergeTypeWithOld);
 
+    if (getLangOpts().HLSL && Context.hasSameFunctionTypeIgnoringParamABI(
```

Comment:
Could you explain a bit more what you're doing here?  At first glance, functions with different out/inout markings aren't compatible.

Reply:
That's correct. We detect and diagnose the mismatch through `mergeParamDeclAttributes`, which gets called by `MergeCompatabileFunctionDecls`, so we need to allow the merge call to proceed even for cases here that might be invalid.

Alternatively we could move that diagnostic out somewhere earlier.

Reply:
Earlier would be more consistent with how other checks like CheckEquivalentExceptionSpec work, I think?

I'm mostly concerned that this logic is twisted in a way that's hard to understand...

Reply:
Can do, will update.